### PR TITLE
Indicate whether routes are provisioned when GETting Alertmanager configuration

### PR DESCRIPTION
**Pull request by alexweav**
**What this PR does / why we need it**:

The GET AM config route will now indicate whether sub-objects are provisioned. This is done by an additional JSON key `provenance`. If an object is not provisioned, the key will not be present.

If the provisioning FF is not enabled, the flag will not be present.

This PR only implements this for notif. policies. It adds the infrastructure for this kind of change though, so the other resources will be much simpler in follow-on PRs.

**Which issue(s) this PR fixes**:

rel: https://github.com/grafana/grafana/issues/36153

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/ngalert/provisioning/notification_policies.go
Comment (alexweav) [inline pkg/services/ngalert/provisioning/notification_policies.go:29]: Drop this inlined thing entirely. Move to `apimodels.Route` which now has all the fields we need.

File: pkg/services/ngalert/provisioning/persist.go
Comment (alexweav) [inline pkg/services/ngalert/provisioning/persist.go:28]: We now had 3 places where this struct was pasted in both impl and tests. Rule of 3 => put it somewhere shared.

I'm not the biggest fan of needing this. Perhaps the provenance store API could be improved to make this not necessary. But, that wasn't the goal of this PR and would require even more files to be touched.

**Comment by joeblubaugh** [comment]
The idea here is to make sure clients don't stomp over "provisioned objects" with manual edits?

File: pkg/services/ngalert/provisioning/persist.go
Comment (joeblubaugh) [inline pkg/services/ngalert/provisioning/persist.go:33]: Do we need the `Resource` prefixes on these functions?

File: pkg/services/ngalert/provisioning/persist.go
Comment (joeblubaugh) [inline pkg/services/ngalert/provisioning/persist.go:29]: Is the usual implementation of this a pointer or struct type?

File: pkg/services/ngalert/provisioning/persist.go
Comment (joeblubaugh) [inline pkg/services/ngalert/provisioning/persist.go:33]: Do you want `a *ProvenanceOrgAdapter`? I'm neutral on the question. How long-lived are these structs?

File: pkg/services/ngalert/api/api_alertmanager_test.go
Comment (joeblubaugh) [inline pkg/services/ngalert/api/api_alertmanager_test.go:223]: Is there any data in the response that needs to be asserted here?

**Review by joeblubaugh — APPROVED** [review]
LGTM - all these comments are optional things you could address, but shouldn't block merging. No need for a follow-up review from me should you choose to take action on any of them.

File: pkg/services/ngalert/api/api_alertmanager_test.go
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/api/api_alertmanager_test.go:223]: Is this test even necessary with the two tests below?

**Review by JohnnyQQQQ — APPROVED** [review]
LGTM - we should add it to the internal changelog, something like: `Alerting: Indicate whether route object is provisioned when GETting Alertmanager configuration` perhaps also changing the PR title like this as it's only the route for now.

File: pkg/services/ngalert/api/api_alertmanager_test.go
Comment (alexweav) [reply pkg/services/ngalert/api/api_alertmanager_test.go:223]: hah, nope, this can be deleted entirely. this was from a draft implementation and is totally surpassed by the tests added below. 🙃

**Comment by alexweav** [comment]
> The idea here is to make sure clients don't stomp over "provisioned objects" with manual edits?

Yes that's the eventual goal of this saga - but this PR doesn't add that enforcement quite yet. Right now we're only making the status visible through the API. This PR lets the UI display a banner if something is provisioned, disable save buttons, etc.

In the near future we'll be adding this validation.

File: pkg/services/ngalert/provisioning/persist.go
Comment (alexweav) [reply pkg/services/ngalert/provisioning/persist.go:33]: Very short lived. They're just simply mapping one interface to another in a few one-off locations.

File: pkg/services/ngalert/provisioning/persist.go
Comment (alexweav) [reply pkg/services/ngalert/provisioning/persist.go:29]: Struct. I'm working on a PR to drop this type entirely, since It's being used in way more places than I expected. Originally, I was expecting this to be a one-off shim that connects two interfaces.

Comment (alexweav) [reply pkg/services/ngalert/provisioning/persist.go:29]: I will keep this how it is, since this object is very short-lived, but I'll end up dropping it.

**Comment by alexweav** [comment]
Modified title to be specific and added to changelog.

File: pkg/services/ngalert/provisioning/persist.go
Comment (alexweav) [reply pkg/services/ngalert/provisioning/persist.go:29]: PR in question: https://github.com/grafana/grafana/pull/48137

that PR deletes this type entirely

**Pull request merged**
